### ISSUE TITLE
Error "Unsequenced modification and access" when building distribution from sources

### ISSUE THREAD

**Issue by gianlucabertani** [google/j2objc#1487]
Hello,

I'm trying to build the distribution from scratch from release 2.7 and Xcode 12.0.1, with the only change to default build parameters being the full list of supported architectures:

```bash
$ export J2OBJC_ARCHS="iphone simulator macosx iphone64 iphone64e watchv7k watch64 watchsimulator simulator64 appletvos appletvsimulator maccatalyst"
$ make dist
```

When it gets to the compilation of Striped64.m I get the following error:

```
compiling /Users/gbertani/j2objc-2.7/jre_emul/build_result/Classes/java/util/concurrent/atomic/Striped64.m
/Users/gbertani/j2objc-2.7/jre_emul/build_result/Classes/java/util/concurrent/atomic/Striped64.m:132:91: error:
      unsequenced modification and access to 'v' [-Werror,-Wunsequenced]
  ...v = JreLoadVolatileLong(&a->value_), (fn == nil) ? v + x : [((id<JavaUtilFunctionLongBinaryOperator>) nil_chk(fn...
       ^                                                ~
/Users/gbertani/j2objc-2.7/jre_emul/build_result/Classes/java/util/concurrent/atomic/Striped64.m:160:82: error:
      unsequenced modification and access to 'v' [-Werror,-Wunsequenced]
  ...v = JreLoadVolatileLong(&base_), (fn == nil) ? v + x : [((id<JavaUtilFunctionLongBinaryOperator>) nil_chk(fn)) a...
       ^                                            ~
2 errors generated.
make[1]: *** [/Users/gbertani/j2objc-2.7/jre_emul/build_result/objs-iphone/java/util/concurrent/atomic/Striped64.o] Error 1
make: *** [jre_emul_dist] Error 2
```

I have read both issues #338 and #343, but none of the suggestions worked:

- First I tried fixing the expression as `(v + x)`, but the error persisted. 
- Then I tried adding `--extract-unsequenced` to `TRANSLATE_ARGS` in `translate.mk`, then cleaning and rebuilding, but the error also persisted.

Any other suggestion?

**Comment by tomball** [comment]
One of the joys of being an Xcode developer is that with each release there's the chances that new clang  are added to what it considers the default set. For Xcode 12, this adds -Wunsequenced. While I generally support enabling all useful warnings, it's still a hassle when they arrive unannounced.

To fix any new warning that is breaking the build, take its name and prepend it with "no-". So in this case, the failing warning is "-Wunsequenced", and the workaround is to specify "-Wno-unsequenced". The j2objc build has a WARNINGS build variable, so to suppress this warning, build using:
```
    $ make -j8 dist WARNINGS=-Wno-unsequenced
```
The fix probably needs to be in the translator, as it looks to be literally translating [an equivalent Java expression](https://github.com/google/j2objc/blob/master/jre_emul/android/platform/libcore/ojluni/src/main/java/java/util/concurrent/atomic/Striped64.java#L257). 

The issue (as I currently understand it) is that Java guarantees that parameter expressions are executed left-to-right ([JLS 15.12.4.2](https://docs.oracle.com/javase/specs/jls/se10/html/jls-15.html#jls-15.12.4.2), but function parameters are not evaluated in a defined order in C/C++/ObjC. I think the right fix is for the translator to pull out the local variable assignment (`v = a.value`), and move it into block with the rest of the expression. That screws up the "else if" chain, though, so the rest of the chain also needs to move into the block (original Java, but this is a translator change):
```
             wasUncontended = true;      // Continue after rehash
-        else if (a.cas(v = a.value,
-                       (fn == null) ? v + x : fn.applyAsLong(v, x)))
-            break;
-        else if (n >= NCPU || cells != as)
-            collide = false;            // At max size or stale
-        else if (!collide)
-            collide = true;
-        else if (cellsBusy == 0 && casCellsBusy()) {
-            try {
-                if (cells == as)        // Expand table unless stale
-                    cells = Arrays.copyOf(as, n << 1);
-            } finally {
-                cellsBusy = 0;
-            collide = false;
-            continue;                   // Retry with expanded table
-        }
+        else {
+            v = a.value;
+            if (a.cas(v,
+                           (fn == null) ? v + x : fn.applyAsLong(v, x)))
+                break;
+            else if (n >= NCPU || cells != as)
+                collide = false;            // At max size or stale
+            else if (!collide)
+                collide = true;
+            else if (cellsBusy == 0 && casCellsBusy()) {
+                try {
+                    if (cells == as)        // Expand table unless stale
+                        cells = Arrays.copyOf(as, n << 1);
+                } finally {
+                    cellsBusy = 0;
+                }
+                collide = false;
+                continue;                   // Retry with expanded table
+             }
+         }
        h = advanceProbe(h);
```
Do you see a simpler change? Replacing the `v`s with `a.value` won't work, since it's possible for a.value to be changed by a different thread:
```
             wasUncontended = true;      // Continue after rehash
-        else if (a.cas(v = a.value,
-                       (fn == null) ? v + x : fn.applyAsLong(v, x)))
+        else if (a.cas(a.value,
+                       (fn == null) ? a.value + x : fn.applyAsLong(a.value, x))) // a.value may have 3 different values
             break;
```

**Comment by tomball** [comment]
Another possibility is to extract the whole boolean expression into a private boolean method:
```
             wasUncontended = true;      // Continue after rehash
-        else if (a.cas(v = a.value,
-                       (fn == null) ? v + x : fn.applyAsLong(v, x)))
+        else if (__f(a, fn, x))
             break;
...
+    private boolean __f(Cell a, LongBinaryOperator fn, long x) {}
+        long v = a.value;
+        return a.cas(v, (fn == null) ? v + x : fn.applyAsLong(v, x));
+    }
```
Hmm, that's starting to look like the same extraction done for capturing lambdas. Have to think about this some more...

**Comment by tomball** [comment]
Much simpler: it's a bug (uncaught edge case) in [UnsequencedExpressionRewriter](https://github.com/google/j2objc/blob/master/translator/src/main/java/com/google/devtools/j2objc/translate/UnsequencedExpressionRewriter.java).

**Comment by gianlucabertani** [comment]
This is a damned hard problem, man. And besides, whoever wrote that code segment in the first place has no idea of what "readable code" means. It took me three full scans of the line to see the assignment of `v`.

What if instead of a property read it was a method call (such as `v = a.next()`) and it was used more than one time in the expression (such as `if (a.cas(v = a.next(), (fn == null) ? v + x : fn.applyAsLong(v = a.next(), x)))`) ? The translator would have to decompose the expression as separate statements.

As a pragmatic guy, I would just send a PR to the Android team with the assignment moved before the test (and a commit message such as `Ever heard of side effects?`), and leave the translator as is.

**Comment by tomball** [comment]
The workaround to build with `-Wno-unsequenced` will enable your build until this is fixed.

FYI, pull request #1488 currently only fixes the pattern found in Striped64.java. With it, the build now fails on another hot mess (err, I mean complex expression 😉) in [android/util/Base64.java](https://github.com/google/j2objc/blob/master/jre_emul/android/frameworks/base/core/java/android/util/Base64.java#L701):
```
    v = (((tailLen > 1 ? tail[t++] : input[p++]) & 0xff) << 10) |
        (((tailLen > 0 ? tail[t++] : input[p++]) & 0xff) << 2);
```
I think what happened in Xcode 12 is that clang's unsequenced warning support was significantly improved. So this is a good thing in the long run, as we need to improve the translator to match their error detection.

**Comment by tomball** [comment]
#1488 has been updated so the project now fully builds with Xcode 12.

**Comment by gianlucabertani** [comment]
Thank you very much for your prompt support!

**Comment by tomball** [comment]
Fixed in #1488.
